perf-cluster-operation/pkg/benchmark.go

```diff
@@ -0,0 +1,142 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pkg
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"path/filepath"
+	"strings"
+
+	yaml "gopkg.in/yaml.v2"
+	"knative.dev/test-infra/shared/common"
+)
+
+const (
+	// clusterConfigFile is the config file needs to be put under the benchmark folder if we want to config
+	// the cluster that runs the benchmark, it must follow the scheme defined as GKECluster here.
+	clusterConfigFile = "cluster.yaml"
+
+	// These default settings will be used for configuring the cluster, if not specified in cluster.yaml.
+	defaultLocation  = "us-central1"
+	defaultNodeCount = 1
+	defaultNodeType  = "n1-standard-4"
+	defaultAddons    = "HorizontalPodAutoscaling,HttpLoadBalancing"
+)
+
+// backupLocations are used in retrying cluster creation, if stockout happens in one location.
+// TODO(chizhg): it's currently not used, use it in the cluster creation retry logic.
+var backupLocations = []string{"us-west1", "us-west2", "us-east1"}
+
+// GKECluster saves the config information for the GKE cluster
+type GKECluster struct {
+	Config ClusterConfig `yaml:"GKECluster,omitempty"`
+}
+
+// ClusterConfig is config for the cluster
+type ClusterConfig struct {
+	Location  string `yaml:"location,omitempty"`
+	NodeCount int64  `yaml:"nodeCount,omitempty"`
+	NodeType  string `yaml:"nodeType,omitempty"`
+	Addons    string `yaml:"addons,omitempty"`
+}
+
+// benchmarkNames returns names of the benchmarks.
+//
+// We put all benchmarks under the benchmarkRoot folder, one subfolder represents one benchmark,
+// here we returns all subfolder names of the root folder.
+func benchmarkNames(benchmarkRoot string) ([]string, error) {
+	names := make([]string, 0)
+	dirs, err := ioutil.ReadDir(benchmarkRoot)
+	if err != nil {
+		return names, fmt.Errorf("failed to list all benchmarks under %q: %v", benchmarkRoot, err)
+	}
+
+	for _, dir := range dirs {
+		names = append(names, dir.Name())
+	}
+	return names, nil
+}
+
+// benchmarkClusters returns the cluster configs for all benchmarks.
+func benchmarkClusters(repo, benchmarkRoot string) (map[string]ClusterConfig, error) {
+	// clusters is a map of cluster configs
+	// key is the cluster name, value is the cluster config
+	clusters := make(map[string]ClusterConfig)
+	benchmarkNames, err := benchmarkNames(benchmarkRoot)
+	if err != nil {
+		return clusters, err
+	}
+
+	for _, benchmarkName := range benchmarkNames {
+		clusterConfig := clusterConfigForBenchmark(benchmarkRoot, benchmarkName)
```

Comment:
Hmm, just curious, is it one of Go best practice to use short var names? I think long vars are more readable.

Reply:
https://github.com/golang/go/wiki/CodeReviewComments#variable-names

```
Variable names in Go should be short rather than long. This is especially true for local variables with limited scope. Prefer c to lineCount. Prefer i to sliceIndex.
```